### PR TITLE
docs(compiler): update ngtsc/ngcc design doc with more details

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/compiler/design/architecture.md [hunk 1]
```diff
@@ -170,6 +170,59 @@ Each of the class decorators can be thought of as class transformers that take t
...
+#### Compiler design
+
+Each "Compiler" which transforms a single decorator into a static field will operate as a "pure function". Given input metadata about a particular type and decorator, it will produce an object describing the field to be added to the type, as well as the initializer value for that field (in Output AST format).
+
+A Compiler must not depend on any inputs not directly passed to it (for example, it must not scan sources or metadata for other symbols). This restriction is important for two reasons:
+
+1. It helps to enforce the Ivy locality principle, since all inputs to the Compiler will be visible.
+2. It protects against incorrect builds during `--watch` mode, since the dependencies between files will be easily traceable.
+
+Compilers will also not take Typescript nodes directly as input, but will operate against information extracted from TS sources by the transformer. In addition to helping enforce the rules above, this restriction also enables Compilers to run in at runtime during JIT mode.
```
Comment (trotyl) [inline packages/compiler/design/architecture.md]: `run in at runtime`?

File: packages/compiler/design/architecture.md [hunk 2]
```diff
@@ -334,6 +455,15 @@ For these reasons, `ngcc` will not use the TS emitter to produce the final patch
...
 
+##### Module systems
+
+The Angular Package Format includes more than one copy of a package's code. At minimum, it includes one ESM5 (ES5 code in ES Modules) entrypoint, one ES2015 entrypoint, and one UMD entrypoint. Some libraries _not_ following the package format may still work in the Angular CLI, if they export code that can be loaded by Webpack.
+
+Thus, `ngcc` will have two approaches for dealing with packages on NPM.
+
+1. APF Path: libraries following the Angular package format will have their source code updated to contain Ivy definitions. This ensures tree-shaking will work properly.
+2. Compatibility Path: libraries where `ngcc` cannot determine how to safely modify the existing code will have a patching operation applied. This patching operation produces a "wrapper" file for each file containing an Angular entity, which re-exports patched versions of the Angular entities. This is not compatible with tree-shaking, but will work for libraries which `ngcc` cannot otherwise understand. A warning will be printed to notify the user they should update the version of the library if possible.
```
Comment (mhevery) [inline packages/compiler/design/architecture.md]: Could you add an example of such a situation?

**Review by mhevery — APPROVED** [review]
Awesome doc! Thanks @alxhub

File: packages/compiler/design/architecture.md (hunk 1 above)
Comment (alxhub) [reply packages/compiler/design/architecture.md]: Fixed.

**Pull request closed without merging**